manual/io/index.md

```diff
@@ -56,784 +46,433 @@ root [1] TFile f1("my.root");
-
-The output prints the date/time, the start record address, the number of bytes in the record, the class name of the record and the compression factor.
+The name of the `TKey` can be either explicitly stated when writing, or it can be determined from `TObject::GetName()` for classes inheriting from `TObject`.
 
 _**Example**_
 
 {% highlight C++ %}
-root[] f.Map()
-
-20191010/122600    At:100     N=114    TFile
-20191010/122600    At:214     N=429    TH1F           CX = 2.31
-20191010/122600    At:643     N=424    TH1F           CX = 2.33
-20191010/122600    At:1067    N=426    TH1F           CX = 2.32
-20191010/122600    At:1493    N=425    TH1F           CX = 2.33
-20191010/122600    At:1918    N=429    TH1F           CX = 2.31
-20191010/122600    At:2347    N=424    TH1F           CX = 2.33
-20191010/122600    At:2771    N=418    TH1F           CX = 2.37
-20191010/122600    At:3189    N=428    TH1F           CX = 2.31
-20191010/122600    At:3617    N=422    TH1F           CX = 2.34
-20191010/122600    At:4039    N=421    TH1F           CX = 2.35
-20191010/122600    At:4460    N=431    TH1F           CX = 2.30
-20191010/122600    At:4891    N=424    TH1F           CX = 2.34
-20191010/122600    At:5315    N=430    TH1F           CX = 2.31
-20191010/122600    At:5745    N=426    TH1F           CX = 2.33
-20191010/122600    At:6171    N=425    TH1F           CX = 2.34
-20191010/122600    At:6596    N=3055   StreamerInfo   CX = 3.08
-20191010/122600    At:9651    N=732    KeysList
-20191010/122600    At:10383   N=53     FreeSegments
-20191010/122600    At:10436   N=1      END
+void WriteHist(TH1* hist) {
+  gDirectory->WriteObject(hist, "explicitName");
+  // OR:
+  hist->Write(); // key named `TH1::GetName()` inside `gDirectory`
+}
 {% endhighlight %}
 
-### Logical contents of a ROOT file
+An object read into memory is independent from the object on disk.
+Changes of the in-memory object are not propagated to disk.
+Instead, a new version of the object needs to be saved, for instance passing `"overwrite"` as option to [ `TDirectory::WriteObject()`](https://root.cern/doc/master/classTDirectory.html#a67b115afae97366254dfd44a7f46f66f), see the documentation of [`TDirectoryFile::WriteTObject()`](https://root.cern/doc/master/classTDirectoryFile.html#ae1bb32dcbb69de7f06a3b5de9d22e852).
 
-ROOT provides not only sequential access to the content of a ROOT file, but also random or direct access.
+Removing an object from a `TDirectory` using [`TDirectory::Delete()`](https://root.cern/doc/master/classTDirectory.html#a09e51150c19fd5a0aa9278a4158202ca) will generally not free the corresponding disk space.
+Instead, the storage occupied by the deleted object will be made available (as a [`TFree`](https://root.cern/doc/master/classTFree.html) for subsequent objects to be written to this file.
+Use → [`hadd`]({{ '/manual/root_files/#merging-root-files-with-hadd' | relative_url }}) to defragment a ROOT file by rewriting it.
```

Comment:
It is clear that `Delete` does not delete the object on disk. It is not clear how to delete the object on disk. Maybe it's not the point of the sentence though

Reply:
Javi's understanding is correct, and it should _not_ be clear that `Delete` does not delete the object on disk, because it does :smile: it just does not free the corresponding disk space. Should we clarify somewhere, @vepadulano ?

Reply:
I think I got distracted by some issue that I remember in the past where `Delete`ing an object from a TFile would not really delete it. As in, running a program that just deletes an object (and does not do anything else with the file, thus it does not fill the space that was previously occupied by that object) and then running another program that lists the content of the same file, would show the previously deleted object. Maybe I just misrember, I haven't tried to reproduce this myself again

Reply:
please open a github issue if that's the case :smiley:

Reply:
Now I remember what was my source of confusion:
```cpp
#include <TFile.h>
#include <TH1.h>

void write(){
    TFile f{"myfile.root", "recreate"};
    TH1F h1{"h1","h1",10,0,10};
    TH1F h2{"h2","h2",10,0,10};
    f.WriteObject(&h1,"");
    f.WriteObject(&h2,"");
}

void read_and_delete(){
    TFile f{"myfile.root","update"};
    f.Delete("h1"); // change to  f.Delete("h1;1") to actually remove the object from the TFile
}

void read_and_list(){
    TFile f{"myfile.root","read"};
    f.ls();
}

int main(){
    write();
    read_and_delete();
    read_and_list();
}
```

In the `read_and_delete` function above, the "h1" histogram is only actually deleted from the TFile if the full namecycle is indicated `h1;1`. If only the name of the histogram is indicated, the `Delete` function deletes it from memory (i.e. for the lifetime of the current open file), but then closing and reopening the file again shows both histograms still there.

Reply:
I'm thinking either clearer documentation in `TDirectory::Delete` or a concrete example here in the manual, or both things

Reply:
It's documented at https://root.cern/doc/master/classTDirectoryFile.html#a24747f020c3276e8a7d2e6c91e5e3f76

Reply:
(and also https://root.cern/doc/master/classTDirectory.html#a09e51150c19fd5a0aa9278a4158202ca )

Reply:
(i'm not saying that the behavior is intuitive, but I don't think there is more to be done here in the manual besides linking to those docs)

Reply:
Yes, fair enough :+1: